### PR TITLE
Replace deepFreeze with Object.freeze()

### LINKED ISSUE #3822: Replace usage of `deep-freeze-strict` dependency
`deep-freeze-strict` is a JavaScript library that provides a way to freeze JavaScript objects and prevent any modifications to them. It works by recursively freezing all nested objects, making them read-only and immutable.

We wish to drop the usage of libraries like deep-freeze-strict in favor of more performant native implementations.

Current usage of `deep-freeze-strict` is in `opensearch_dashboards_utils` plugin in `src/plugins/opensearch_dashboards_utils/common/state_containers/create_state_container.ts`. 

```
import deepFreeze from 'deep-freeze-strict';

const defaultFreeze: <T>(value: T) => T = isProduction
  ? <T>(value: T) => value as T
  : <T>(value: T): T => {
      const isFreezable = value !== null && typeof value === 'object';
      if (isFreezable) return deepFreeze(value) as T;
      return value as T;
    };
```

This `defaultFreeze` function is used to freeze states. When creating a state container, if no other `freeze()` function is provided, then the `defaultFreeze()` will be used as default.

### PR DESCRIPTION
**Pull request by sayuree**
### Description
Replaced deepFreeze to native Object.freeze() by recursively freezing the properties of an object

### Issues Resolved
closes #3822 

### Check List

- [ ] All tests pass
  - [x] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [x] New functionality includes testing.
- [ ] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -20,7 +20,7 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - [Security] Bumps hapi/statehood to 7.0.4 ([#3411](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3411))
 - [CVE-2023-25166] Bump formula to 3.0.1 ([#3416](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3416))
 - [CVE-2023-25653] Bump node-jose to 2.2.0 ([#3445](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3445))
-- [CVE-2023-26486][CVE-2023-26487] Bump vega from 5.22.1 to 5.23.0 ([#3533](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3533))
+- [CVE-2023-26486][cve-2023-26487] Bump vega from 5.22.1 to 5.23.0 ([#3533](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3533))
```
Comment (AMoo-Miki) [inline CHANGELOG.md:23]: I don't think you intended to make this change. Please revert this one.

File: src/plugins/opensearch_dashboards_utils/common/state_containers/create_state_container.ts [hunk 2]
```diff
@@ -47,12 +46,18 @@ const isProduction =
...
     : !process.env.NODE_ENV || process.env.NODE_ENV === 'production';
 
+const deepFreeze = (value: any) => {
+  if (value && typeof value === 'object' && !Object.isFrozen(value)) {
+    Object.freeze(value);
+    Object.getOwnPropertyNames(value).forEach((prop) => deepFreeze(value[prop]));
+  }
+  return value;
+};
+
```
Comment (AMoo-Miki) [inline src/plugins/opensearch_dashboards_utils/common/state_containers/create_state_container.ts:56]: `deep-freeze-strict` handled freezing functions. Will we not be needing that?

Comment (sayuree) [reply src/plugins/opensearch_dashboards_utils/common/state_containers/create_state_container.ts]: @AMoo-Miki I added new changes, reflecting freezing functions.

**Review by manasvinibs — COMMENTED** [review]
Changes looks good to me. Also I think the intent of the issue #3822 is to get rid of the dependency itself. Do you want to include cleaning up the dependency `deep-freeze-strict` as its not required anymore with your changes.

**Comment by sayuree** [comment]
@manasvinibs Oh, I will remove the dependency itself. Thanks!

**Comment by sayuree** [comment]
> @manasvinibs Oh, I will remove the dependency itself. Thanks!

I have removed extra dependency using **yarn remove**, however, each time I want to commit changes I keep getting the following error:
image
I have checked PATH and everything seems fine: I am still able to make commits for other repos.

**Comment by abbyhu2000** [comment]
> > @manasvinibs Oh, I will remove the dependency itself. Thanks!
> 
> I have removed extra dependency using **yarn remove**, however, each time I want to commit changes I keep getting the following error: image I have checked PATH and everything seems fine: I am still able to make commits for other repos.

You prob need to check your environment variables. Could you do `nvm which node`? it should match the PATH.

File: src/plugins/opensearch_dashboards_utils/common/state_containers/create_state_container.ts [hunk 3]
```diff
@@ -47,12 +46,22 @@ const isProduction =
     ? process.env.NODE_ENV === 'production'
     : !process.env.NODE_ENV || process.env.NODE_ENV === 'production';
 
+const deepFreeze = (value: any) => {
```
Comment (ananzh) [inline src/plugins/opensearch_dashboards_utils/common/state_containers/create_state_container.ts:49]: We have a deepFreeze definition in this utility package, packages/osd-std/src/deep_freeze.ts:
```
import { RecursiveReadonly } from '@osd/utility-types';

/** @public */
export type Freezable = { [k: string]: any } | any[];

/**
 * Apply Object.freeze to a value recursively and convert the return type to
 * Readonly variant recursively
 *
 * @public
 */
export function deepFreeze<T extends Freezable>(object: T) {
  // for any properties that reference an object, makes sure that object is
  // recursively frozen as well
  for (const value of Object.values(object)) {
    if (value !== null && typeof value === 'object') {
      deepFreeze(value);
    }
  }
  return Object.freeze(object) as RecursiveReadonly<T>;
}
```
Could we move this one in in file as well? The difference I could see is that this one  includes function freezing. So maybe we modify deepFreeze in this file as a basic freeze function. Then extend it to include function freezing? Therefore, other files can use it and code is cleaner. What do you think?

Comment (AMoo-Miki) [reply src/plugins/opensearch_dashboards_utils/common/state_containers/create_state_container.ts:49]: I agree with Anan. We should port yours to replace the one linked above as yours is better on a bunch of levels. Then import `deepFreeze` from there.

**Comment by sayuree** [comment]
> @sayuree do you need more help on this PR?

I apologize for not answering quickly. Can I finish this PR myself?

**Comment by joshuarrrr** [comment]
> > @sayuree do you need more help on this PR?
> 
> I apologize for not answering quickly. Can I finish this PR myself?

Sure! Thanks for the response.

**Comment by abbyhu2000** [comment]
@sayuree how is the progress on this PR? Will convert it to draft for now, u can mark it as ready once u have the updates. Let us know if u have any questions or need help, thanks!

**Comment by ananzh** [comment]
@sayuree Due to an extended period without updates, we're going to close this issue for now. We appreciate your contribution and understand that priorities and availability can change. Please feel free to reopen this issue when you have the opportunity to revisit it. Thank you for your understanding.

**Pull request closed without merging**